### ISSUE TITLE
If a folder contains .acl files will they get copied after `copyFolder` call?

### ISSUE THREAD

**Issue by aorumbayev** [jeff-zucker/solid-file-client#38]
This is more of a general question, rather than an issue. 
I have tested a copyFolder call with several .json and .ttl files and it seems like .acl files describing access control to those .json and .ttl files were not copied. 
My questions here is, how do we generally treat the .acl files - do they have to be unique per each file and after performing the copyFolder, I need to loop over new copies and regenerate .acl files for them? This approach would seem rather redundant.

**Comment by jeff-zucker** [comment]
Great question.  A new version of copyFolder is coming soon and will copy .acl files if requested.  If you'd like to see this in operation, try recursiveCopy in my fork of rdflib at  https://github.com/jeff-zucker/rdflib.js.  My plan is to finish testing the fork and then merge the fork back into the main rdflib and then include a call to it in solid-file-client which will then support copying .acl files as part of the copyFolder.  For now, yes, the only option is to specifically copy each .acl.  It is even more of a pain in the neck with deleting because if you delete the file but not the .acl you can't then delete the folder.  Aiming to fix that with recursiveDelete which will by default delete .acl files.  It is also in my rdflib fork.

**Comment by aorumbayev** [comment]
@jeff-zucker just wanted to ask if there are there any updates on the this feature ? :) 
I am working on a Master's thesis and majority of work related to solid, one of the features I am currently implementing is related to copying of folder and I rely heavily on .acl files inside them. 

Therefore, if there is any help needed for testing/merging the forked repo - I will be glad to contribute, just let me know :)

**Comment by jeff-zucker** [comment]
Thanks for the nudge to act :-).   I've discovered an issue with my handling of the link header for ACL files and now have a much more robust solution.  See https://forum.solidproject.org/t/solid-recursive-copy/1119/13.  I've now updated the recursiveCopy in my fork and am testing the changes  Could you try out my fork using the aclCopy:true flag on some folders that have ACL files at multiple levels? You can use the example/recursiveOps/ files to test.  I'm moving this to the top of my priority list so hope to have it merged soon.

**Comment by jeff-zucker** [comment]
There was a typo in the example, if you downloaded before now, download again. The correct flag to get the ACL files copied is fetcher.recursiveCopy( url, {copyACL:true})

**Comment by aorumbayev** [comment]
Hello @jeff-zucker , 
So quick updates - I've tested out the recusriveCopy with latest your commits with copyACL flag, 

it seems like it worked fine and copied all files including acl files on multiple levels. Here is the log below : 

``` bash
node browserless-copy.js https://aorumbayev2.solid.community/linkedpipes https://aorumbayev2.solid.community/lin
kedpipes1
logging in
logged in as <https://aorumbayev2.solid.community/profile/card#me>
Copying folder linkedpipes1
  into https://aorumbayev2.solid.community/

Copying file <https://aorumbayev2.solid.community/linkedpipes/configurations/1558015194002.476.json>
  to <https://aorumbayev2.solid.community/linkedpipes1/configurations/1558015194002.476.json>

Copying file <https://aorumbayev2.solid.community/linkedpipes/configurations/1558015194002.476.json.ttl>
  to <https://aorumbayev2.solid.community/linkedpipes1/configurations/1558015194002.476.json.ttl>

Copying folder configurations
  into https://aorumbayev2.solid.community/linkedpipes1/

Copying ACL https://aorumbayev2.solid.community/linkedpipes/.acl
  to https://aorumbayev2.solid.community/linkedpipes1/.acl

Copying ACL https://aorumbayev2.solid.community/linkedpipes/configurations/1558015194002.476.json.acl
  to https://aorumbayev2.solid.community/linkedpipes1/configurations/1558015194002.476.json.acl

Copying ACL https://aorumbayev2.solid.community/linkedpipes/configurations/1558015194002.476.json.ttl.acl
  to https://aorumbayev2.solid.community/linkedpipes1/configurations/1558015194002.476.json.ttl.acl

Success! 
```
 
If u look at the `/linkedpipes` folder, there is a `/configurations` folder inside that contains some .json and .ttl files each defined with .acl file. The `/configurations` folder also contains .acl file. All of that was copied properly into `/linkedpipes1`, however, when I try to open specific .json files inside copied `/configurations` folder I get `No permission to access this resource` error. I assume something got wrong within the copied .acl files ?

**Comment by aorumbayev** [comment]
@jeff-zucker it seems like only folder .acl files are being copied properly. 
Here is what I get in the sharing settings tab in solid pane when I try to expand the .json file in there 

```
Error reading default ACL. status Can't get headers to find ACL for <https://aorumbayev2.solid.community/linkedpipes1/configurations/1558015194002.476.json>: Failed to load <https://aorumbayev2.solid.community/linkedpipes1/configurations/1558015194002.476.json> Fetcher: <https://aorumbayev2.solid.community/linkedpipes1/configurations/1558015194002.476.json> Forbidden status: 403: undefined
```

**Comment by jeff-zucker** [comment]
Hmm.  A couple of questions - 1) what filetype are the files listed as in the newly created folder? 2) Does this happen only with json files or other files too?  3) Are you able to look at the newly created .acl files to see if their contents are as expected?

**Comment by aorumbayev** [comment]
1) So the folders are all marked as `Basic Container` and files are `Basic Resource`, there is one `.json` and one `.ttl` file inside the `/configurations` folder.
2) It happens with `.ttl` files as well but I assume in that case any file will have same behaviour. 
3) No I'm not - when I try to open .acl files for files I also get the `No permission to access this resource`. I can however look into the .acl files of copied folders and they are identical to original ones.


One additional thing I have noticed is, if I disable specific sharing for those 2 files via solid pane, copying works fine and files are visible in resulting folder (I assume since no file specific .acl files were created). Here is the result of copying same exact folder to a folder without specific .acl files per files - https://aorumbayev2.solid.community/linkedpipes9/

**Comment by jeff-zucker** [comment]
Could you paste in one of the actual .acl files.

**Comment by aorumbayev** [comment]
Sure here is the .acl file used for one of the .json files

```
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#owner> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/auth/acl#Authorization> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#owner> <http://www.w3.org/ns/auth/acl#accessTo> <https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#owner> <http://www.w3.org/ns/auth/acl#agent> <https://aorumbayev2.solid.community/profile/card#me> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#owner> <http://www.w3.org/ns/auth/acl#mode> <http://www.w3.org/ns/auth/acl#Control> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#owner> <http://www.w3.org/ns/auth/acl#mode> <http://www.w3.org/ns/auth/acl#Read> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#owner> <http://www.w3.org/ns/auth/acl#mode> <http://www.w3.org/ns/auth/acl#Write> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#public> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/auth/acl#Authorization> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#public> <http://www.w3.org/ns/auth/acl#accessTo> <https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#public> <http://www.w3.org/ns/auth/acl#agentClass> <http://xmlns.com/foaf/0.1/Agent> .
<https://aorumbayev2.solid.community/linkedpipes/configurations/1558018063706.1357.json.acl#public> <http://www.w3.org/ns/auth/acl#mode> <http://www.w3.org/ns/auth/acl#Read> .
```

**Comment by jeff-zucker** [comment]
That's the original .acl file?  With all URLs using absolute addresses?  Or that's some sort of dump from the file?  For comparison, here's what one of my .acl files looks like:
```
@prefix : <#>.
@prefix n0: <http://www.w3.org/ns/auth/acl#>.
@prefix c: </profile/card#>.
@prefix n1: <http://xmlns.com/foaf/0.1/>.
:owner n0:accessTo <foo.ttl>; n0:agent c:me; n0:mode n0:Control, n0:Read, n0:Write.
:public n0:accessTo <foo.ttl>; n0:agentClass n1:Agent; n0:mode n0:Read.
``` 
Note that there are no absolute URLs.  If your .acl files do contain absolute URLs, recursiveCopy() will not be able to handle them, it only copies, does not modify contents of the files.  If this is indeed the case that you absolute URLs in the original .acl files, could you try changing them to relative and see if that works?

**Comment by aorumbayev** [comment]
No that's not the original one since I generate and update the ACL files programmatically. @jeff-zucker thanks for pointing out , I'll have to debug my function that generates the .ACL files. But yes I think that might be issue. 

Aside from that, the copy function itself works perfectly 🙂👍

**Comment by jeff-zucker** [comment]
Thank you for testing.  Please let me know if you are able to get it working by altering the .acl files.  I wonder how the folder .acl files are different.  BTW, [Solid IDE](https://jeff-zucker.github.io/solid-ide/) is a useful way to look at the actual .acl file.  You have to use the location bar to specificalyl ask for e.g. https://me.example/public/.acl to look at the main folder .acl.

**Comment by aorumbayev** [comment]
@jeff-zucker - some more updates:

I have checked out the code I have been using for .acl file generation (mostly based of code from [Pixolid](https://github.com/carloss8/pixolid)). The .acl files were just a concatenation of `Statement` objects converted to string, I've parsed that into a store and then serialised it back to Turtle, that got rid of absolute paths. 

So to sum up, here are some additional logs of recursive copy with .acl files on multiple levels, everything got copied properly and works fine. 

```
node browserless-copy.js https://aorumbayev2.solid.community/linkedpipes70 https://aorumbayev2.solid.community/linkedpipes100
logging in
logged in as <https://aorumbayev2.solid.community/profile/card#me>
Copying folder linkedpipes100
  into https://aorumbayev2.solid.community/

Copying file <https://aorumbayev2.solid.community/linkedpipes70/configurations/1558060795261.2612.json>
  to <https://aorumbayev2.solid.community/linkedpipes100/configurations/1558060795261.2612.json>

Copying file <https://aorumbayev2.solid.community/linkedpipes70/configurations/1558060795261.2612.json.ttl>
  to <https://aorumbayev2.solid.community/linkedpipes100/configurations/1558060795261.2612.json.ttl>

Copying ACL https://aorumbayev2.solid.community/linkedpipes70/.acl
  to https://aorumbayev2.solid.community/linkedpipes100/.acl

Copying folder configurations
  into https://aorumbayev2.solid.community/linkedpipes100/

Copying ACL https://aorumbayev2.solid.community/linkedpipes70/configurations/1558060795261.2612.json.acl
  to https://aorumbayev2.solid.community/linkedpipes100/configurations/1558060795261.2612.json.acl

Copying ACL https://aorumbayev2.solid.community/linkedpipes70/configurations/1558060795261.2612.json.ttl.acl
  to https://aorumbayev2.solid.community/linkedpipes100/configurations/1558060795261.2612.json.ttl.acl

Copying ACL https://aorumbayev2.solid.community/linkedpipes70/configurations/.acl
  to https://aorumbayev2.solid.community/linkedpipes100/configurations/.acl

Success!
```

And here are samples of .acl files for folders and files:
• File : 
```
@prefix : <#>.
@prefix n0: <http://www.w3.org/ns/auth/acl#>.
@prefix c: </profile/card#>.
@prefix n1: <http://xmlns.com/foaf/0.1/>.

:owner
    a n0:Authorization;
    n0:accessTo <1558060795261.2612.json>;
    n0:agent c:me;
    n0:mode n0:Control, n0:Read, n0:Write.
:public
    a n0:Authorization;
    n0:accessTo <1558060795261.2612.json>;
    n0:agentClass n1:Agent;
    n0:mode n0:Read.
```

• Folder : 
```
@prefix : <#>.
@prefix n0: <http://www.w3.org/ns/auth/acl#>.
@prefix con: <./>.
@prefix c: </profile/card#>.
@prefix n1: <http://xmlns.com/foaf/0.1/>.

:owner
    a n0:Authorization;
    n0:accessTo con:;
    n0:agent c:me;
    n0:defaultForNew con:;
    n0:mode n0:Control, n0:Read, n0:Write.
:public
    a n0:Authorization;
    n0:accessTo con:;
    n0:agentClass n1:Agent;
    n0:defaultForNew con:;
    n0:mode n0:Read.
```

The only case when the copying crashes for me is if some of the folders along the level has no .acl file defined . It produces some weird promise rejection error. But I assume the main SOLID requirement on that is that the developer is responsible for creating .acl file for any file/folder creating within solid pod ?
Here is the error the call generates whenever some folder has no .acl : 
```
(node:16838) UnhandledPromiseRejectionWarning: Error: Web error: 403 (Forbidden) on PUT of <https://aorumbayev2.solid.community/linkedpipes102/configurations/1558061060013.058.json.ttl.acl>
    at /Users/aorumbayev/Thesis/examples/rdflib.js/lib/fetcher.js:1410:22
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:16838) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:16838) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:16838) UnhandledPromiseRejectionWarning: Error: Web error: 403 (Forbidden) on PUT of <https://aorumbayev2.solid.community/linkedpipes102/configurations/1558061060013.058.json.acl>
    at /Users/aorumbayev/Thesis/examples/rdflib.js/lib/fetcher.js:1410:22
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:16838) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
```

**Comment by jeff-zucker** [comment]
Phew, I'm glad that got rid of the authorization errors.  It's probably good practice to keep links relative anyway.  I've added a warning to the readme about it.  Solid itself does not require you to create an .acl file for every resource.  But the only way for an app to know if there is an .acl is to try to read the .acl so therefore there are errors from the GET request for the non-existent file. It's something the app should handle, so it's on me to find where the unhandled exceptions are and catch them before they get  to the user.  In this case they're harmless because the correct files get created anyway but it's better for me to clean them up so thanks for finding them!

**Comment by jeff-zucker** [comment]
Fixed in version 1.0.0.
